### PR TITLE
build: make HAVE_O_CLOEXEC available outside LevelDB (bugfix)

### PR DESCRIPTION
**Pull request by theStack**
The result of the O_CLOEXEC availability check is currently only set in the Makefile and passed to LevelDB (see `LEVELDB_CPPFLAGS_INT` in `src/Makefile.leveldb.include`), but not defined to be used in our codebase. This means that code within the preprocessor conditional `#if HAVE_O_CLOEXEC` was actually never compiled. On the master branch this is currently used for pipe creation in `src/shutdown.cpp`, PR #21007 moves this part to a new module (I found the issue while testing that PR).

The fix is similar to the one in #19803, which solved the same problem for HAVE_FDATASYNC.

In the course of working on the PR it turned out that pipe2 is not available an all platforms, hence a configure check and a corresponding define HAVE_PIPE2 is introduced and used.

The PR can be tested by anyone with a system that has pipe2 and O_CLOEXEC available by putting gibberish into the HAVE_O_CLOEXEC block: on master, everything should compile fine, on PR, the compiler should abort with an error. At least that's my naive way of testing preprocessor logic, happy to hear more sophisticated ways :-)

### CODE REVIEW

**Comment by fanquake** [comment]
Both of the macOS builds are failing with:
```bash
  CXX      libbitcoin_server_a-shutdown.o
shutdown.cpp:38:9: error: use of undeclared identifier 'pipe2'
    if (pipe2(g_shutdown_pipe, O_CLOEXEC) != 0) {
        ^
1 error generated.
make[2]: *** [Makefile:8763: libbitcoin_server_a-shutdown.o] Error 1
```

Looks like it has `O_CLOEXEC` but not pipe2.

**Comment by theStack** [comment]
Added a commit to change the preprocessor condition to only use popen2 if O_CLOEXEC is available _and_ MacOS is not used.
Also, re-opened the PR to start Cirrus again (the failure was unrelated to this PR, see #21216).

**Review by luke-jr — CHANGES_REQUESTED** [review]
At the very least, have configure check specifically for `pipe2`...

Ideally, a code path that sets `O_CLOEXEC` after the normal `pipe` path could be nice too.

File: configure.ac [hunk 1]
```diff
@@ -1149,6 +1149,16 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <fcntl.h>]],
...
+AC_DEFINE_UNQUOTED([HAVE_O_CLOEXEC], [$HAVE_O_CLOEXEC], [Define to 1 if O_CLOEXEC flag is available.])
+
+AC_MSG_CHECKING(for pipe2)
+AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <unistd.h>]],
+ [[ static int pipefd[2];
+    pipe2(pipefd, 0); ]])],
+ [ AC_MSG_RESULT(yes); HAVE_PIPE2=1 ],
+ [ AC_MSG_RESULT(no); HAVE_PIPE2=0 ]
+)
+AC_DEFINE_UNQUOTED([HAVE_PIPE2], [$HAVE_PIPE2], [Define to 1 if pipe2 is available.])
```
Comment (luke-jr) [inline configure.ac:1161]:
```suggestion
AC_CHECK_DECLS([pipe2])
```

**Comment by laanwj** [comment]
> This means that code within the preprocessor conditional #if HAVE_O_CLOEXEC was actually never compiled

I'm confused—how does that compile at all if the preprocessor flag is mot set? Our reason for using `#if X` insteadof `#ifdef X` is to prevent issues like this. But apparently that doesn't work?

File: configure.ac (hunk 1 above)
Comment (theStack) [reply configure.ac]: Heh, nice that sometimes the solution can be so simple, thanks. Any reason why the same approach is not used for also e.g. `fdatasync`? (That's what I used as template in the previous version).

**Comment by theStack** [comment]
@laanwj: Yes, I would have also expected `#if` to fail if a symbol wasn't declared at all, but apparently that's not the case :/

Force-pushed now with a much simpler pipe2 detection in configure, [as suggested by luke-jr](https://github.com/bitcoin/bitcoin/pull/21250#discussion_r579890630).

**Review by luke-jr — APPROVED** [review]
utACK

**Comment by laanwj** [comment]
Code review ACK 9bac71350d98580cc7441957fc7c3fa2f4158553

> @laanwj: Yes, I would have also expected #if to fail if a symbol wasn't declared at all, but apparently that's not the case :/

I feel kind of ashamed for not realizing this for so long. I wonder if we can come up with some new convention that does ensure this! (anything I can think of inside `#if` would become kind of long…) and document that in the developer notes.

Not in this PR, of course.

**Pull request merged**